src/metabase/query_processor/pivot.clj

```diff
@@ -1,98 +1,144 @@
+(defn- generate-queries
   "Generate the additional queries to perform a generic pivot table"
-  [request]
-  (let [query                (:query request)
-        all-breakouts        (:breakout query)
-        bitmask-index        (map-invert (into {} (map-indexed hash-map all-breakouts)))
-        new-breakouts        (generate-specified-breakouts all-breakouts
-                                                           (:pivot_rows request)
-                                                           (:pivot_cols request))
-        bitwise-not-truncate (fn [num-bits val]
-                               (- (dec (bit-shift-left 1 num-bits)) val))]
-    (map (fn [breakout]
-           ;; this implements basically what PostgreSQL does for grouping -
-           ;; look at the original set of groups - if that column is part of *this*
-           ;; group, then set the appropriate bit (entry 1 sets bit 1, etc)
-           ;; at the end, perform a bitwise-not with a mask. Doing it manually
-           ;; because (bit-not) extends to a Long, and twos-complement and such
-           ;; make that messy.
-           (let [start-mask (reduce
-                             #(bit-or (bit-shift-left 1 %2) %1) 0
-                             (map #(get bitmask-index %) breakout))]
-             (add-grouping-field request breakout
-                                 (bitwise-not-truncate (count all-breakouts) start-mask))))
-         new-breakouts)))
+  [{{all-breakouts :breakout} :query, :keys [pivot-rows pivot-cols query], :as outer-query}]
+  (try
+    (for [breakout-indecies (breakout-combinations (count all-breakouts) pivot-rows pivot-cols)
+          :let              [group-number    (group-number (count all-breakouts) breakout-indecies)
+                             new-breakouts (for [i breakout-indecies]
+                                             (nth all-breakouts i))]]
+      (add-grouping-field outer-query new-breakouts group-number))
+    (catch Throwable e
+      (throw (ex-info (tru "Error generating pivot queries")
+                      {:type qp.error-type/qp, :query query}
+                      e)))))
+
+;; this function needs to be executed at the start of every new query to
+;; determine the mapping for maintaining query shape
+(defn- map-cols
+  {:arglists '([query context])}
+  [query {all-expected-cols ::all-expected-columns}]
+  (assert (seq all-expected-cols))
+  (let [query-cols (map-indexed vector (qp/query->expected-cols query))]
+    (map (fn [item]
+           (some #(when (= (:name item) (:name (second %)))
+                    (first %))
+                 query-cols))
+         all-expected-cols)))
+
+;; this function needs to be called for each row so that it can actually
+;; shape the row according to the `map-cols` fn above
+(defn- map-row
+  [row cols]
+  ;; the first query doesn't need any special mapping, it already has all the columns
+  (if cols
+    (for [i cols]
+      (when i
+        (nth row i)))
+    row))
```

Comment:
I moved this stuff out of `run-pivot-query` so it could be closer to where it's actually used